### PR TITLE
Order addItem endpoint implements MQTT

### PR DESCRIPTION
**Pull request by pietro99**
I made a proof of concept of how to implement the MQTT communication.
I implemented it for the addItem enpoint

- stock service subscribes to /subtractStock messages
https://github.com/mirunabetianu/WSM-Project/blob/540b416a816e97a149f600a3e9a86835e43f508d/stock/main.go#L37
- order service subscribes to /response messages
https://github.com/mirunabetianu/WSM-Project/blob/540b416a816e97a149f600a3e9a86835e43f508d/order/main.go#L48
- http call to /addItem happens
- order service publishes a /subtractStock message to the broker with item id, amount to remove (alway 1) and a unique id. then waits for a new /response message or time-out
https://github.com/mirunabetianu/WSM-Project/blob/540b416a816e97a149f600a3e9a86835e43f508d/order/main.go#L154-L166
- stock service receives the /subtractStock and removes stock and publishes a /response messge with the message id and the response
- order service receives the /response message, ends the loop in the request and returns the response of the stock service

### CODE REVIEW

File: order/main.go [hunk 1]
```diff
@@ -119,7 +151,20 @@ func addItemToOrder(c *fiber.Ctx) error {
...
+			var payload string = itemId + "/1/" + request_id
+			utils.Publish(payload, mqttClient, "topic/subtractStock")
+
+			//wait for response
+			start := time.Now()
+			for !responses[request_id].has_arrived {
+				if time.Since(start) >= WAIT_TIME {
+					return c.SendStatus(408)
+				}
+				time.Sleep(1 * time.Nanosecond) // to avoid "concurrent map read and map write" error
```
Comment (Pepijnk12) [inline order/main.go:165]: Could this potentially still give an error when the read/write takes longer than 1 nanosecond?

Comment (pietro99) [reply order/main.go:165]: I am not sure about this or how to test it. but I have tried to implement map as a sync.Map which would be a solution but I was not able to make it work.

File: order/utils/mqtt.go [hunk 2]
```diff
@@ -43,18 +44,21 @@ var connectLostHandler mqtt.ConnectionLostHandler = func(client mqtt.Client, err
...
-	}
+func Publish(payload interface{}, client mqtt.Client, topic string) {
+	fmt.Printf("\npublishing to topic: %v with payload: %v\n", topic, payload)
+	token := client.Publish(topic, 0, true, payload)
+	token.Wait()
+	//time.Sleep(time.Second)
 }
 
 func Subscribe(client mqtt.Client, topic string) {
 	token := client.Subscribe(topic, 1, nil)
```
Comment (Mozafar7) [inline order/utils/mqtt.go:55]: Shouldn't there be some error handling involved here? Something like: 
```
if token.Error() != nil
 do something with the error
```

File: order/utils/mqtt.go [hunk 3]
```diff
@@ -43,18 +44,21 @@ var connectLostHandler mqtt.ConnectionLostHandler = func(client mqtt.Client, err
...
 	token.Wait()
```
Comment (Mozafar7) [inline order/utils/mqtt.go:56]: Shouldn't there be some error handling involved here (after this line)? Something like: 
```
if token.Error() != nil
 do something with the error
```

**Comment by farsene** [comment]
We can close this now as Miruna implemented all the endpoints with go routines

**Pull request closed without merging**